### ISSUE TITLE
Framework adapters (e.g. Svelte)

### ISSUE THREAD

**Issue by Blakintosh** [shadcn-ui/lint#12]
Hey there, love the direction of this project. 

Are there any plans to make it easier for the community to add support for frameworks such as Svelte, for example, by exposing an adapter API? From reading the source, template analysis appears to target JSX/TSX, while Svelte uses its own template syntax.

Is something like this planned, or would separate projects for different frameworks make more sense?

Possibly related to #2.

**Comment by SquidwardHill** [comment]
Seconding this! (We use shadcn-svelte)

**Comment by Misophistful** [comment]
I tried this on a SvelteKit + Tailwind v4 + shadcn-svelte codebase, 231 `.svelte` files. Two things that might be useful for the adapter design.

## The current failure mode is silent

Run the rules over a `.svelte` file with `svelte-eslint-parser` and you get zero problems and exit 0. The same markup in `.tsx`:

```
Probe.tsx      4 errors  (bg-red-50, text-red-800, p-[13px], inline style)
Probe.svelte   0 errors  (same classes, same rules enabled)
```

`CallExpression` is already framework-agnostic, so a `cva()` or `tv()` call inside a `<script>` block does get checked while the markup beside it doesn't. That half-state reads as "configured and passing" rather than "not supported". For an agent-facing linter that's the expensive kind of wrong. A warning when a file produced no class sites would help on its own, whatever happens with adapters.

## The collector is closer to framework-agnostic than it looks

I expected `sites/collect.ts` to need a rewrite. It doesn't. The value walker in `collectClassStrings` is already a generic ESTree walker: `ConditionalExpression`, `LogicalExpression`, `ObjectExpression`, `TemplateLiteral` gluing, identifier resolution, the `cva` and `tv` config shapes. One branch in it is JSX-specific, `case "JSXExpressionContainer"`, three lines long.

Five node types differ. Four are structurally identical:

| JSX                                                   | Svelte                                            |
| ----------------------------------------------------- | ------------------------------------------------- |
| `JSXAttribute`, `.name.name`                          | `SvelteAttribute`, `.key.name`                    |
| `JSXExpressionContainer` to `.expression`             | `SvelteMustacheTag` to `.expression`              |
| `Literal`, `.value`                                   | `SvelteLiteral`, `.value`                         |
| `JSXElement` ancestor walk via `.openingElement.name` | `SvelteElement`, names itself via `.name`         |
| `JSXMemberExpression` `{object, property}`            | `SvelteMemberExpressionName` `{object, property}` |

Two real differences past the renaming.

An attribute value is an array rather than a node. `class="static {cond ? 'a' : ''}"` parses as `[SvelteLiteral, SvelteMustacheTag]`. Synthesizing a `TemplateLiteral` would keep your glued-interpolation handling. I approximated with `ArrayExpression` and lost it.

`project/parser.ts` can't open a component file. Variant axes live in a `.svelte` file's `<script>` block, so `extractVariantDefinitions` needs that block pulled out first. Padding it with spaces keeps every range pointing at the right offset.

I patched the built `dist/index.js` as a prototype, 10 hunks, +45/-9 lines. `no-raw-colors`, `no-arbitrary-values` and `no-restyle` then all worked on Svelte markup, variant suggestions included, read out of a `tv()` call:

```
"p-4" is not allowed on <RailRow>: <RailRow> owns its spacing. Use a size
(sm, md, lg), or margin here or gap on the parent for space around it.
```

On the real codebase that gave 94 findings outside the vendored `ui/` directory in 6.7 seconds, off an unmodified shadcn-svelte `components.json`. I didn't test `no-unknown-classes` or `require-static-classes`.

## The one structural problem

Every `no-restyle` finding carried the right file and the wrong component name. `<default>`, or `<//\n\tRoot>`.

That isn't a patch bug. In JSX, `export function Button` carries the name. In Svelte every component is a file's default export, and shadcn-svelte's barrels re-export it as `Root` or `default`:

```ts
import Root from "./label.svelte";
export {
  Root,
  //
  Root as Label,
};
```

The resolved export name carries nothing, so the message says `<default> owns its typography` where a React user sees `<CardTitle>`. The usage site does have the name the author wrote, `<Card.Title>` or `<Label.Root>`. Preferring it would change JSX behaviour too, where resolution over name is deliberate: an import into the ui directory is that component whatever it got renamed to.

So which way do you want it? The display name could fall back to the usage-site name when the resolved export is `default`. Or an adapter could supply its own name resolver. I lean towards the second, though it's a much bigger surface.

**Comment by DrJKL** [comment]
A vote for Vue too.

**Comment by stickerdaniel** [comment]
Same experiment on my shadcn-svelte app, and I land on @Misophistful's mapping. Two class and style channels it doesn't cover.

`style:background-color={color}` never becomes a `style` attribute, so `no-inline-styles` reports nothing. Not hypothetical: shadcn-svelte's registry ships two of these itself. `SvelteStyleDirective` holds its value in the same shape as `SvelteAttribute`, longform `(SvelteLiteral | SvelteMustacheTagText)[]`, shorthand `[]` with the key carrying the name.

`class:is-exit={cond}` is a class site with no `class` attribute. `SvelteClassDirective`, key is the class name, shorthand leaves `expression: null`. 1 use in my `src`, 0 in the registry, so the volume is low and the file still passes.

`SvelteElement.kind` is `"html"`, `"component"` or `"special"`. The native/component split that costs an import resolution in JSX is on the node here. `"special"` is `<svelte:element this={tag}>` and needs a decision.

**Comment by rorybush** [comment]
@Misophistful I checked the `// Root` naming issue and it also happens with a plain TSX component. Here's the small example I tried.

`label.tsx`:

```tsx
export default function Root() {
  return null;
}
```

`index.ts`:

```ts
import Root from "./label.tsx";

export {
  Root,
  //
  Root as Label,
};
```

Both `exportsOf()` and `definingExportOf()` resolve `Label` to the barrel file, with the comment included in the binding name. Removing the comment gives the correct component file and `Root`.

`parseNamed()` in `packages/lint/src/project/modules.ts` splits the export list without handling comments, so it treats the comment as part of the identifier. That particular bug exists independently of Svelte support. The `<default>` naming case is a separate question.

One other detail that might help with the adapter work: explicit `./label.svelte` imports already resolve. The resolver accepts an existing file before trying its extension list, so the missing `.svelte` entry doesn't block those imports. Parsing the component still needs separate support.
